XenAdmin/Dialogs/UpgradeVm/UpgradeVmDialog.cs

```diff
@@ -0,0 +1,469 @@
+                    case VmUpgradeState.UPDATEABLE: return Messages.UPGRADE_VMS_STATE_UPDATEABLE;
+                }
+
+                return status.ToString();
+            }
+
+            private string GetActionStringForVm()
+            {
+                var status = VMStateHelper.GetUpgradeStateForVm(VM);
+
+                if (status == VmUpgradeState.UPGRADING_GO_TO_CONSOLE)
+                    return Messages.UPGRADE_VMS_ACTIONS_GO_TO_CONSOLE;
+
+                if (status == VmUpgradeState.ERROR)
+                    return Messages.UPGRADE_VMS_ACTIONS_ERROR;
+
+                return string.Empty;
+            }
+
+            private bool CanBeUpgraded()
+            {
+                var status = VMStateHelper.GetUpgradeStateForVm(VM);
+                
+                return CAN_UPGRADE_STATES.Contains(status);
+            }
+        }
+
+        private enum VmUpgradeState 
+        {
+            UNKNOWN = 0,
+            ERROR = 1,
+            UPGRADABLE = 2,
+            INSTALL_MANAGEMENT_AGENT = 3,
+            UPGRADING = 4,
+            UPGRADING_GO_TO_CONSOLE = 5,
+            UPGRADED = 6,
+            INITIALIZING = 7,
+            NOT_SHOWN = 8,
+            UPDATEABLE = 9,
+        }
+
+        private static class VMStateHelper
+        {
+            internal static VmUpgradeState GetUpgradeStateForVm(VM vm)
+            {
+                //if (!vm.is_a_real_vm || !vm.IsWindows || ???)
+                //    return VmUpgradeState.NOT_SHOWN;
+                
+                var guestToClient = vm.GuestToClientValue;
+                var clientToGuest = vm.ClientToGuestValue;
+
+                if (clientToGuest == VM.ClientToGuestMessage.EMPTY && guestToClient == VM.GuestToClientMessage.EMPTY)
+                    return VmUpgradeState.UPGRADABLE;
+
+                if (clientToGuest == VM.ClientToGuestMessage.INITIALIZING)
+                    return VmUpgradeState.INITIALIZING;
+
+                if (clientToGuest == VM.ClientToGuestMessage.GOTOCONSOLE)
+                    return VmUpgradeState.UPGRADING_GO_TO_CONSOLE;
+
+                if (clientToGuest == VM.ClientToGuestMessage.UPGRADING
+                    || guestToClient == VM.GuestToClientMessage.STARTED
+                    || guestToClient == VM.GuestToClientMessage.RUNNING
+                    )
+                    return VmUpgradeState.UPGRADING;
+
+                if (guestToClient == VM.GuestToClientMessage.DONE)
+                    return VmUpgradeState.UPGRADED;
+
+                if (guestToClient == VM.GuestToClientMessage.ERROR)
+                    return VmUpgradeState.ERROR;
+
+                //if (mgmnt agent running, version < version_Dundee) //case for old management agent
+                //    return VmUpgradeState.UPDATEABLE;
+
+                return VmUpgradeState.UNKNOWN;
+            }
+        }
+
+        private void ShowVmConsole(VM vm)
+        {
+            mainWindow.Invoke(delegate
+            {

```

Comment:
Code review comment: we don't think Invoke is needed here, Gabor needs to check this